cmd/ctl/pkg/upgrade/migrate/migrate.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return err
+	}
+	if !allV1 {
+		fmt.Fprintln(o.ErrOut, "It looks like you are running a pre-1.0 version of cert-manager. Please upgrade cert-manager to v1.6 before upgrading to v1.7.")
+		return fmt.Errorf("migration failed")
+	}
+	fmt.Fprintln(o.Out, "All CustomResourceDefinitions have 'v1' configured as the storage version.")
+
+	/*
+		fmt.Fprintln(o.Out, "Looking for CRDs that contain resources that require migrating to 'v1'...")
+		crdsRequiringMigration, err := o.discoverCRDsRequiringMigration(ctx, "v1", allCRDNames)
+		if err != nil {
+			fmt.Fprintf(o.ErrOut, "Failed to determine resource types that require migration: %v\n", err)
+			return err
+		}
+		if len(crdsRequiringMigration) == 0 {
+			fmt.Fprintln(o.Out, "Nothing to do. cert-manager CRDs do not have 'status.storedVersions' containing old API versions. You may proceed to upgrade to cert-manager v1.7.")
+			return nil
+		}
+	*/
+
+	fmt.Fprintf(o.Out, "Found %d resource types that require migration:\n", len(crdsRequiringMigration))
+	for _, crd := range crdsRequiringMigration {
+		fmt.Fprintf(o.Out, " - %s\n", crd.Name)
+	}
+
+	for _, crd := range crdsRequiringMigration {
+		fmt.Fprintf(o.Out, "Migrating %q objects in group %q - this may take a while...\n", crd.Spec.Names.Kind, crd.Spec.Group)
+
+		if err := o.migrateResourcesForCRD(ctx, crd); err != nil {
+			fmt.Fprintf(o.ErrOut, "Failed to migrate resource: %v\n", err)
+			return err
+		}
+	}
+
+	fmt.Fprintln(o.Out, "Patching CRD resources to set 'status.storedVersions' to 'v1'...")
+	if err := o.patchCRDStoredVersions(ctx, crdsRequiringMigration); err != nil {
+		fmt.Fprintf(o.ErrOut, "Failed to patch 'status.storedVersions' field: %v\n", err)
+		return err
+	}
+
+	fmt.Fprintln(o.Out, "Successfully migrated all cert-manager resource types. It is now safe to proceed with upgrading to cert-manager v1.7.")
+	return nil
+}
+
+func (o *Options) ensureCRDStorageVersionEquals(ctx context.Context, vers string, names []string) (bool, []*apiext.CustomResourceDefinition, error) {
+	var crds []*apiext.CustomResourceDefinition
+	for _, crdName := range names {
+		crd := &apiext.CustomResourceDefinition{}
+		if err := o.client.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
+			return false, nil, err
+		}
+
+		// Discover the storage version
+		storageVersion := ""
+		for _, v := range crd.Spec.Versions {
+			if v.Storage {
+				storageVersion = v.Name
+				break
+			}
+		}
+
+		if storageVersion != vers {
+			fmt.Fprintf(o.Out, "CustomResourceDefinition object %q has storage version set to %q. You MUST upgrade to cert-manager v1.0-v1.6 before migrating resources for v1.7.\n", crdName, storageVersion)
+			return false, nil, nil
+		}
+
+		crds = append(crds, crd)
+	}
+
+	return true, crds, nil
+}
+
+func (o *Options) discoverCRDsRequiringMigration(ctx context.Context, desiredStorageVersion string, names []string) ([]*apiext.CustomResourceDefinition, error) {
```

Comment:
I would vote for always running the full migration and removing this bit, especially since we package this as a CLI command, so users explicitly choose to run it